src/rgw/driver/sfs/writer.cc

```
@@ -38,13 +43,109 @@ SFSAtomicWriter::SFSAtomicWriter(
+                       << ec << dendl;
+    switch (ec.value()) {
+      case ENOSPC:
+        return -ERR_QUOTA_EXCEEDED;
+      default:
+        return -ERR_INTERNAL_ERROR;
+    }
+  }
+
+  int ret;
+  ret = ::open(object_path.c_str(), O_CREAT | O_TRUNC | O_CLOEXEC | O_WRONLY,
+               0644);
+  if (ret < 0) {
+    lsfs_dout(dpp, 10) << "error opening file " << object_path << ": "
+                       << cpp_strerror(-fd) << dendl;
+    return -ERR_INTERNAL_ERROR;
+  }
+
+  fd = ret;
+  return 0;
+}
+
+int SFSAtomicWriter::close() noexcept {
+  int result = 0;
+  int ret;
+
+  ret = ::fsync(fd);
+  if (ret < 0) {
+    lsfs_dout(dpp, 10) << fmt::format("failed to fsync fd:{}: {}. continuing.",
+                                      fd, cpp_strerror(ret))
+                       << dendl;
+  }
+
+  ret = ::close(fd);
+  fd = -1;
+  if (ret < 0) {
+    lsfs_dout(dpp, 10) << fmt::format("failed closing fd:{}: {}. continuing.",
+                                      fd, cpp_strerror(ret))
+                       << dendl;
+    switch (ret) {
+      case -EDQUOT:
+      case -ENOSPC:
+        result = -ERR_QUOTA_EXCEEDED;
+        break;
+      default:
+        result = -ERR_INTERNAL_ERROR;
+    }
+    io_failed = true;
+  }
+  return result;
+}
+
+void SFSAtomicWriter::cleanup() noexcept {
+  lsfs_dout(dpp, 10)
+      << fmt::format(
+             "cleaning up failed upload to file {}. "
+             "returning error.",
+             object_path.string())
+      << dendl;
+
+  std::error_code ec;
+  std::filesystem::remove(object_path, ec);
+  if (ec) {
+    lsfs_dout(dpp, 10) << fmt::format("failed deleting file {}: {} {}. ignoring.",
+                                      object_path.string(), ec.message(), ec.value())
+                       << dendl;
+  }
+
+  const auto dir_fd = ::open(object_path.parent_path().c_str(), O_RDONLY);
+  int ret = ::fsync(dir_fd);
+  if (ret < 0) {
+    lsfs_dout(dpp, 10)
```

if we're logging failures, might be better to log them to level `-1`. Would be nice to have defined an `lsfs_derr()` macro, but given that's not available, logging to `-1` would suffice. Otherwise these errors will be hidden most of the time in a production environment.

---

:+1: